qt/widgets/common/src/WorkspaceMultiSelector.cpp

```diff
@@ -0,0 +1,433 @@
+    if (*maximum > 0)
+      return QString("0-%1").arg(*maximum);
+    return "0";
+  }
+  return "";
+}
+
+QString getIndexString(const std::string &workspaceName) { return getIndexString(getWorkspace(workspaceName)); }
+
+std::unique_ptr<QRegExpValidator> createValidator(const QString &regex, QObject *parent) {
+  return std::make_unique<QRegExpValidator>(QRegExp(regex), parent);
+}
+
+QString OR(const QString &lhs, const QString &rhs) { return "(" + lhs + "|" + rhs + ")"; }
+
+QString NATURAL_NUMBER(std::size_t digits) { return OR("0", "[1-9][0-9]{," + QString::number(digits - 1) + "}"); }
+
+const QString EMPTY = "^$";
+const QString SPACE = "(\\s)*";
+const QString COMMA = SPACE + "," + SPACE;
+const QString MINUS = "\\-";
+
+const QString NUMBER = NATURAL_NUMBER(4);
+const QString NATURAL_RANGE = "(" + NUMBER + MINUS + NUMBER + ")";
+const QString NATURAL_OR_RANGE = OR(NATURAL_RANGE, NUMBER);
+const QString SPECTRA_LIST = "(" + NATURAL_OR_RANGE + "(" + COMMA + NATURAL_OR_RANGE + ")*)";
+
+class SpectraListDelegate : public QStyledItemDelegate {
+public:
+  QWidget *createEditor(QWidget *parent, const QStyleOptionViewItem & /*option*/,
+                        const QModelIndex & /*index*/) const override {
+    auto lineEdit = std::make_unique<QLineEdit>(parent);
+    auto validator = createValidator(SPECTRA_LIST, parent);
+    lineEdit->setValidator(validator.release());
+    return lineEdit.release();
+  }
+
+  void setEditorData(QWidget *editor, const QModelIndex &index) const override {
+    const auto value = index.model()->data(index, Qt::EditRole).toString();
+    static_cast<QLineEdit *>(editor)->setText(value);
+  }
+
+  void setModelData(QWidget *editor, QAbstractItemModel *model, const QModelIndex &index) const override {
+    auto *lineEdit = static_cast<QLineEdit *>(editor);
+    model->setData(index, lineEdit->text(), Qt::EditRole);
+  }
+
+  void updateEditorGeometry(QWidget *editor, const QStyleOptionViewItem &option,
+                            const QModelIndex & /*index*/) const override {
+    editor->setGeometry(option.rect);
+  }
+};
+} // namespace
+
+/**
+ * Default constructor
+ * @param parent :: A widget to act as this widget's parent (default = NULL)
+ * @param init :: If true then the widget will make calls to the framework
+ * (default = true)
+ */
+namespace MantidQt {
+namespace MantidWidgets {
+
+WorkspaceMultiSelector::WorkspaceMultiSelector(QWidget *parent, bool init)
+    : QTableWidget(parent), m_addObserver(*this, &WorkspaceMultiSelector::handleAddEvent),
+      m_remObserver(*this, &WorkspaceMultiSelector::handleRemEvent),
+      m_clearObserver(*this, &WorkspaceMultiSelector::handleClearEvent),
+      m_renameObserver(*this, &WorkspaceMultiSelector::handleRenameEvent),
+      m_replaceObserver(*this, &WorkspaceMultiSelector::handleReplaceEvent), m_init(init), m_connected(false),
+      m_workspaceTypes(), m_showHidden(false), m_showGroups(false), m_binLimits(std::make_pair(0, -1)), m_suffix() {
```

Comment:
Because the bin limits are initially empty, maybe we could use a std::optional and instantiate it as a std::nullopt until the bin limits are assigned?